ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,862 @@
+        hints_bag
+    }
+}
+
+/// Compute commitments for UncheckedLeaf
+pub fn compute_commitments(leaf: UncheckedLeaf) -> Option<FirstDlogProverMessage> {
+    let mut ret: Option<FirstDlogProverMessage> = None;
+
+    match leaf {
+        UncheckedLeaf::UncheckedSchnorr(pdlog) => {
+            let challenge = pdlog.challenge;
+            let proposition = pdlog.proposition;
+            let second_message = pdlog.second_message;
+            let comm = compute_commitment(&proposition, &challenge, &second_message);
+            ret = Some(FirstDlogProverMessage::from(comm));
+        }
+        UncheckedLeaf::UncheckedDhTuple(_) => {}
+    }
+    ret
+}
+
+#[allow(clippy::unwrap_used)]
+/// Outputs Hints bag for a signed(invalid) transaction
+pub fn bag_for_multi_sig(
+    sigma_tree: SigmaBoolean,
+    real_propositions: &[SigmaBoolean],
+    simulated_propositions: &[SigmaBoolean],
+    proof: &[u8],
+) -> HintsBag {
+    let ut: UncheckedTree = parse_sig_compute_challenges(&sigma_tree, proof.to_owned()).unwrap();
+    let mut bag: HintsBag = HintsBag::empty();
+    traverse_node(
+        ut,
+        real_propositions,
+        simulated_propositions,
+        NodePosition::crypto_tree_prefix(),
+        &mut bag,
+    );
+    bag
+}
+
+/// A method which is generating commitments to randomness. A commitment is about a first step
+/// of a zero-knowledge proof-of-knowledge knowledge protocol.
+/// the commitments are generated for every input box of transaction
+/// and return as `TransactionHintsBag`.
+/// generated commitments corresponds to `public_keys` that prepared in function input
+pub fn generate_commitments(
+    tx_context: TransactionContext,
+    state_context: &ErgoStateContext,
+    public_keys: &[SigmaBoolean],
+) -> Result<TransactionHintsBag, TxSigningError> {
+    let tx = tx_context.spending_tx.clone();
+    let mut hints_bag = TransactionHintsBag::empty();
+    for (i, input) in tx.inputs.iter().enumerate() {
+        let input_box = tx_context
+            .get_boxes_to_spend()
+            .find(|b| b.box_id() == input.box_id)
+            .ok_or(TxSigningError::InputBoxNotFound(i))?;
+        let ctx = Rc::new(make_context(state_context, &tx_context, i)?);
+        let tree = input_box.ergo_tree.clone();
+        let exp = tree
+            .proposition()
+            .map_err(ProverError::ErgoTreeError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let reduction_result = reduce_to_crypto(&exp, &Env::empty(), ctx)
+            .map_err(ProverError::EvalError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+
+        let sigma_tree = reduction_result.sigma_prop;
+        hints_bag.add_hints_for_input(i, generate_commitments_for(sigma_tree, public_keys));
+    }
+    Ok(hints_bag)
+}
+
+/// Extracting hints from a transaction and outputs it's corresponding TransactionHintsBag
+pub fn extract_hints(
+    signed_tx: Transaction,
+    tx_context: TransactionContext,
```

Comment:
Since `TransactionContext` holds an unsigned tx which has no place here let's make `TransactionContext<T>` polymorphic in `spending_tx` type to allow both signed and unsigned transactions. Thus, eliminating the need for `signed_tx` parameter here. 
Hint: To make it polymorphic you're going to need a new `ErgoTransaction` trait to make `TransactionContext::new` work.